doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,633 @@
+QUIC Connection State Machine
+=============================
+
+FSM Model
+---------
+
+QUIC client-side connection state can be broken down into five coarse phases of
+a QUIC connection:
+
+- The Idle substate (which is simply the state before we have started trying to
+  establish a connection);
+- The Active state, which comprises two substates:
+    - The Establishing state, which comprises many different substates;
+    - The Open state;
+- The Terminating state, which comprises several substates;
+- The Terminated state, which is the terminal state.
+
+There is monotonic progression through these phases.
+
+These names have been deliberately chosen to use different terminology to common
+QUIC terms such as 'handshake' to avoid confusion, as they are not the same
+concepts. For example, the Establishing state uses Initial, Handshake and 1-RTT
+packets.
+
+This discussion is (currently) given from the client side perspective only.
+State machine considerations only relevant to servers are not mentioned.
+0-RTT is also not currently modelled in this analysis.
+
+The synthesis of this FSM is not suggested by the QUIC RFCs but has been
+discerned from the requirements imposed. This does not mean that the
+implementation of this FSM as literally presented below is an optimal or
+advisable implementation strategy, and a cursory examination of existing QUIC
+implementations suggests that such an approach is not common. Moreover, excess
+attention should not be given to the Open state, as 1-RTT application
+communication can occur even still in the Establishing state (for example, when
+the handshake has been completed but not yet confirmed).
+
+However, the state machine described herein is helpful as an aid to
+understanding and broadly captures the logic which our implementation will
+intend to embody. The design of the actual implementation is discussed further
+below.
+
+The above states and their substates are defined as follows:
+
+- The Establishing state involves the use of Initial and Handshake
+  packets. It is terminated when the handshake is confirmed.
+
+  Handshake confirmation is not the same as handshake completion.
+  Handshake confirmation occurs on the client when it receives
+  a `HANDSHAKE_DONE` frame (which occurs in a 1-RTT packet, thus
+  1-RTT packets are also invoked in the Establishing state).
+  On the server, handshake confirmation occurs as soon as
+  the handshake is considered completed.
```

Comment:
Any chance of a more precision description of when?
I guess not critical for MVP.

Reply:
These terms are defined in RFC 9001 s. 4.1. Added reference.